sale_order_type/models/sale.py

```diff
@@ -73,12 +74,27 @@ def onchange_type_id(self):
 
     @api.model
     def create(self, vals):
+        if self:
+            # This is a copy
+            # Give it a generic name
+            vals["name"] = "Copy of %s" % self[0].name
+            # Force user to select type_id, so proper
+            # sequencing will be maintained
+            vals["type_id"] = False
         if vals.get("name", "/") == "/" and vals.get("type_id"):
-            sale_type = self.env["sale.order.type"].browse(vals["type_id"])
-            if sale_type.sequence_id:
-                vals["name"] = sale_type.sequence_id.next_by_id()
+            vals["name"] = self._get_sequencing(vals["type_id"]) or vals["name"]
         return super(SaleOrder, self).create(vals)
 
+    def write(self, vals):
+        """
+        Maintain proper sequencing when copying
+
+        """
+        type_id = vals.get("type_id")
+        if type_id:
+            vals["name"] = self._get_sequencing(type_id) or self.name
+        super(SaleOrder, self).write(vals)
```

Comment:
I gather this is for when the `type_id` changes and the number needs to change as well. That's cool but, this looks like it will get incremented every time the `type_id` is written, which is not what you want.

Also, if records in `self` are many, then many sale orders will get the same sequence, which is also not good.

You can get this right by, before `super`, looping the records "for this in self", compare their type_id with what it's going to be, and if you see a change, add it to a {so: new_type_id, so: new_type_id} style dict.

On the remaining sale orders (self - dict.keys()), you call super like you call it now, but without vals["name"].

Afterwards, for each one that had their type changed, you add the new sequence to vals and call super(SaleOrder, this).write(vals).

That way you won't slow down regular writes by adding a loop when it's not needed.

Reply:
Yes you are right, I should multi this and your suggestion is absolutely the way to do it.